feat: Implement Actor Model pattern with automatic actor ID and communication with loose coupling #3251Actor model

ssrijan-007-sys opened the pull request:
## Description

This PR enhances the implementation of the **Actor Model** pattern in Java.

### What’s included:
- Automatic ID assignment to actors (e.g., actor-1, actor-2)
- Multi-actor interaction with message passing
- Ansh and Srijan actors demonstrate safe communication
- Fixes infinite message loop using reply limit logic
- `README.md` and `etc/` are updated

Closes #3251

## Checklist ✅

- [x] Code follows [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html)
- [x] Code formatted using `./mvnw spotless:apply`
- [x] Includes:
  - [x] `src/`
  - [x] `README.md`
  - [x] `etc/` diagram
  - [x] `pom.xml`
- [x] Module added to parent `pom.xml`
- [x] Code builds with `./mvnw clean install`

Looking forward to your feedback!

ssrijan-007-sys commented:
✅ Implementation Completed
Implemented the Actor Model pattern with automatic actor ID and loose-coupling message communication.

Added unit tests for all key classes and verified actual message passing logic.

Increased code coverage to 93.9% (from 45.5%).

Fixed Spotless formatting violations.

Build & tests pass locally via mvn clean install and mvn test.

Quality Gate passed on SonarCloud.

Please review the PR when you get a chance. I'm happy to make further adjustments if needed!

iluwatar commented on actor-model/README.md:
```diff
@@ -0,0 +1,52 @@
+# Actor Model
```
Check out the `README.md` template here: https://github.com/iluwatar/java-design-patterns/wiki/21.-New-Pattern-Template

iluwatar commented on actor-model/pom.xml:
```diff
@@ -0,0 +1,63 @@
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <version>5.10.0</version>
```
Use the version from the parent pom.xml

iluwatar commented on actor-model/pom.xml:
```diff
@@ -0,0 +1,63 @@
+            <version>5.10.0</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
```
Add maven-assembly-plugin so that we can execute the jar from the command line

iluwatar commented on actor-model/src/main/java/com/iluwatar/actormodel/App.java:
```diff
@@ -0,0 +1,43 @@
+ */
+package com.iluwatar.actormodel;
+
+public class App {
```
Here, above the `App` class, describe the actor model. What are the main components and interactions? Add commentary how this example implements the pattern.

iluwatar commented on actor-model/src/main/java/com/iluwatar/actormodel/ExampleActor.java:
```diff
@@ -0,0 +1,47 @@
+    this.actorSystem = actorSystem;
+  }
+
+  Logger logger = Logger.getLogger(getClass().getName());
```
Use Lombok's @Slf4j annotation for logging

iluwatar commented on actor-model/src/main/java/com/iluwatar/actormodel/Message.java:
```diff
@@ -0,0 +1,43 @@
+  private final String content;
+  private final String senderId;
+
+  public Message(String content, String senderId) {
+    this.content = content;
+    this.senderId = senderId;
+  }
```
Use Lombok's @AllArgsConstructor

iluwatar commented on actor-model/src/main/java/com/iluwatar/actormodel/Message.java:
```diff
@@ -0,0 +1,43 @@
+    this.senderId = senderId;
+  }
+
+  public String getContent() {
+    return content;
+  }
+
+  public String getSenderId() {
+    return senderId;
+  }
```
Use @Getter

iluwatar commented on actor-model/src/test/java/com/iluwatar/actor/ActorModelTest.java:
```diff
@@ -0,0 +1,58 @@
+
+    ansh.send(new Message("Hello Srijan", srijan.getActorId()));
+
+    // You can improve this later by capturing output or using mocks
+    assertNotNull(srijan.getActorId());
```
It would be nice to implement the message passing test properly

ssrijan-007-sys commented:
✅ Requested Changes Implemented

Refactored logging using Lombok’s @Slf4j annotation

Implemented proper message-passing verification in the unit test (ActorModelTest)

Finalized and committed a fully structured README.md following the project format:

Includes real-world example, diagrams, benefits/trade-offs, usage, etc.

Added working links to references, tutorials, and related patterns

🔁 Kindly review the updates and let me know if any additional changes are needed. Happy to iterate further if required!

iluwatar commented:
Looks good! Thank you for the contribution 🎉

@all-contributors please add @ssrijan-007-sys for code
